### PR TITLE
Updates to set VM name, storage type & IP allocation type

### PR DESCRIPTION
**Pull request by andiradulescu**
Hey @ecalder6 !

Really love your work!

I needed some changes to your scripts so I took the liberty to add everything I needed.

Please take a look and let me know if you need me to make any changes in order to merge this.

Cheers,
Andrei

P.S. You can test what I did here: https://github.com/andiradulescu/azure-gaming/ ("dev" branch)

---

* Added VmName
* Added VmStorageType ("StandardSSD_LRS" for Standard SSD, "Standard_LRS" for Standard HDD)
* Added VmIpType ("Static", "Dynamic")
* Removed Standard_NV6_Promo from LowPri since it's not supported

---
EDIT: the following were removed from this PR:
* Switched to Windows 10 image (20h1-pro)
* Install ZeroTier only when networkID is present
* Updated the Tesla M60 driver

### CODE REVIEW

**Comment by ecalder6** [comment]
Thanks for the PR! The problem with using Windows 10 is that it requires a volume license.
image

**Comment by andiradulescu** [comment]
I tried to look up more info related to what you say, but with not much success. Can you explain or give me a link to where you found what’s in the screenshot?

From what I believe, Microsoft should be charging the Windows Pro license automatically, since the image is provided by Microsoft, and it doesn’t “complain” (it appears as “Activated”).

When you create a new VM from the Azure portal, you have the option to select “Windows 10 Pro”. This makes me, as a user, believe the pricing for the license is going to be “pay as you go”, as the cost for the VM is also “pay as you go”.
image

In configuration, after the VM is created, this is shown.
image

What I can do next is contact Azure and ask if I need a license or it’s being charged automatically, to clarify this.

**Comment by AmineI** [comment]
Interesting changes, besides the Windows 10 choice. What benefits does it brings, by the way ?

> I tried to look up more info related to what you say, but with not much success. Can you explain or give me a link to where you found what’s in the screenshot?

There's no link I'm afraid since this is a dynamic page, but you can access it by typing "Windows 10" in Azure's search bar. You can also see #15 for previous talks about that

**Comment by ecalder6** [comment]
Windows 10 more consumer friendly than Windows Server. One major benefit of using Windows 10 is that you can play Windows Store games (including Xbox Game Pass for PC).

I agree that these licensing info are hard to find. Before making the change to Windows 10, we need to be 100% sure that all users (not just users with enterprise/volume license) can deploy Windows 10 machine.

Below is what you see when you try to create a windows 10 VM. It will be difficult to convince Microsoft to change their Windows licensing policy. 
image

**Comment by AmineI** [comment]
> Windows 10 is more consumer friendly than Windows Server. One major benefit of using Windows 10 is that you can play Windows Store games (including Xbox Game Pass for PC).
Ohhh, right, the Windows Store ! I totally labeled this as a "Azure-*steam*-gaming" thingy, so the consumer friendly part did not matter because it'd always be in Steam. But you're right, the store would be a significant benefit !

**Comment by andiradulescu** [comment]
@AmineI - besides Windows 10 the benefits are:
- removed Standard_NV6_Promo from LowPri since it's not supported - it works with Standard_NV6
- option to set the VmName - in my scenario I didn't want "CloudGaming" as the VM name and I couldn't deploy two VMs with this template since they had the same name
- option to set IP allocation to Static/Dynamic - I noticed I got logged out of Steam when using a Dynamic IP, after restarting the VM - also, it was annoying because I also need to access the VM from mobile with RD Client and I needed to set the new IP after every restart
- option to set storage type to Standard SSD besides Standard HDD - I know @ecalder6 isn't a fan :) of Standard SSD, but I did a benchmark and SSD is somewhat faster on read speed

All of these were needed in my own scenario. I can change this PR as you guys see fit. @AmineI I really like what you did in https://github.com/ecalder6/azure-gaming/pull/41 - @ecalder6 +1 to merge it - after getting merged, I will also update my PR.

Also, I opened a service request with Azure and I expect an answer regarding the licensing. I got a reply that from the Support Professional that he "will be reaching out to my Windows 10 team" and get back to me.

**Comment by andiradulescu** [comment]
Standard HDD/SSD benchmark results:
Standard HDD
Standard SSD

**Comment by AmineI** [comment]
> @AmineI - besides Windows 10 the benefits are:
> 
> * removed Standard_NV6_Promo from LowPri since it's not supported - it works with Standard_NV6
> * option to set the VmName - in my scenario I didn't want "CloudGaming" as the VM name and I couldn't deploy two VMs with this template since they had the same name
> * option to set IP allocation to Static/Dynamic - I noticed I got logged out of Steam when using a Dynamic IP, after restarting the VM - also, it was annoying because I also need to access the VM from mobile with RD Client and I needed to set the new IP after every restart
> * option to set storage type to Standard SSD besides Standard HDD - I know @ecalder6 isn't a fan :) of Standard SSD, but I did a benchmark and SSD is somewhat faster on read speed
> 
> All of these were needed in my own scenario. I can change this PR as you guys see fit. @AmineI I really like what you did in #41 - @ecalder6 +1 to merge it - after getting merged, I will also update my PR.
> 
> Also, I opened a service request with Azure and I expect an answer regarding the licensing. I got a reply from the Support Professional that he "will be reaching out to Windows 10 team" and get back to me.

Yes, I noticed and liked your changes as well - I didn't mean to say your PR had no benefit, on the contrary ! I was only wondering about the benefits of switching to W10, not the others. 
I would have updated to Windows Server 2019 and not W10, but the Windows Store advantage makes sense ! The only dark point is the Windows 10 licensing issue

**Comment by ecalder6** [comment]
> * option to set storage type to Standard SSD besides Standard HDD - I know @ecalder6 isn't a fan :) of Standard SSD, but I did a benchmark and SSD is somewhat faster on read speed

I am a fan of SSD :)

The problem is that I don't think real SSDs are supported for NV skus. [This doc](https://docs.microsoft.com/en-us/azure/virtual-machines/nv-series) mentions that [Premium Storage](https://azure.microsoft.com/en-us/blog/introducing-premium-storage-high-performance-storage-for-azure-virtual-machine-workloads/) is not supported for NV series VMs.

I believe only NVv2 series and later support SSD, as noted by this [bullet point](https://docs.microsoft.com/en-us/azure/virtual-machines/sizes-gpu#deployment-considerations):
> N-series VMs differ in the type of Azure Storage they support for their disks. NC and NV VMs only support VM disks that are backed by Standard Disk Storage (HDD). NCv2, NCv3, ND, NDv2, and NVv2 VMs only support VM disks that are backed by Premium Disk Storage (SSD).

**Comment by AmineI** [comment]
> The problem is that I don't think real SSDs are supported for NV skus. [This doc](https://docs.microsoft.com/en-us/azure/virtual-machines/nv-series) mentions that [Premium Storage](https://azure.microsoft.com/en-us/blog/introducing-premium-storage-high-performance-storage-for-azure-virtual-machine-workloads/) is not supported for NV series VMs.

Correct, but Standard SSD should work on [any VM](https://azure.microsoft.com/fr-fr/blog/preview-standard-ssd-disks-for-azure-virtual-machine-workloads/), which is already a benefit over Standard HDD, although lower than Premium SSD

**Comment by andiradulescu** [comment]
@ecalder6 - I was joking about you not being a fan of "Azure Standard SSD" - I'm sure you're a fan of SSDs as everyone else ;)

More details about Standard SSDs here: https://azure.microsoft.com/en-gb/blog/preview-standard-ssd-disks-for-azure-virtual-machine-workloads/

**_Performance**

Standard SSD disks are designed to provide single-digit millisecond latencies for most IO operations. Also, the above disk sizes are designed to deliver up to 500 IOPS and 60 MBPS throughput similar to the HDD disks. Actual IOPS and Throughput may vary sometimes depending on the traffic patterns. Standard SSD disks will provide more consistent performance than the HDD disks with the lower latency.

Premium SSD disks, on the other hand, perform better than Standard SSD disks, with very low latencies, high IOPS/throughput and even better consistency with provisioned disk performance, and it is the recommended disk type for all other production workloads.

Like the existing Premium SSD Disks, Standard SSD Disks also use IO Unit size of 256KB. If the data being transferred is less than 256 KB, it is considered 1 I/O unit. Larger I/O sizes are counted as multiple I/Os of size 256 KB. For example, a 1,100 KB I/O is counted as five I/O units._

**Comment by ecalder6** [comment]
Got it! I didn't know standard SSD was a thing. Thanks for correcting me! 

Lmk when Microsoft responds. Also feel free to split Windows 10 change to another PR if you'd like to check this one in first.

**Comment by AmineI** [comment]
> Also feel free to split Windows 10 change to another PR if you'd like to check this one in first.

Perhaps the best of both worlds would indeed be a separate PR, implementing a parameter for the OS. There are a few examples on this [Azure blog post](https://azure.microsoft.com/en-us/blog/create-flexible-arm-templates-using-conditions-and-logical-functions/). 

That would let people choose between W10/WS, and thus get rid of any eventual licensing issues (Only users that own the appropriate licences would select W10, the other would have to select WS). But perhaps it would introduce configuration issues - I'm unsure if all the configuration done in the extensions/setup would work on both WS & W10.

**Comment by andiradulescu** [comment]
@AmineI @ecalder6 - cleaned the changes included in this PR - let me know what I need to add - changes to README, maybe something else.

**Comment by AmineI** [comment]
> Cleaned the changes included in this PR - let me know what I need to add - changes to README, maybe something else.

Just reviewed indeed ! I think adding the IP type description to the Readme (in the description of the parameters) could be a nice addition, especially since they could get charged if you have more than 5, according to https://azure.microsoft.com/en-us/pricing/details/ip-addresses/ . 

"CloudGaming" is hard-written in the readme for quite a few instructions, we could replace it by "The VM Name", but it is a minor change.

**Comment by ecalder6** [comment]
Changes look good! Should be good to merge after you rebase.

File: Standard.json [hunk 1]
```diff
@@ -23,7 +30,30 @@
             ],
             "metadata": {
                 "description": "The type of the VM to deploy. Note that only NV-series support gaming."
-            }
+            },
+            "defaultValue": "Standard_NV6_Promo"
+        },
+        "VmStorageType": {
```
Comment (ecalder6) [inline Standard.json:36]: Add description for these new parameters in readme.

**Comment by andiradulescu** [comment]
@ecalder6 @AmineI - I updated the README and also added Premium SSD to the template. Let me know if anything else is needed. Thanks!

**Pull request merged**